### ISSUE TITLE
valid-jsdoc fails on functions that use ES6 destructuring assignment in signature

### ISSUE THREAD

**Issue by lencioni** [eslint/eslint#2051]
Given the following code:

```js
/**
 * @param {object} options
 * @return {number}
 */
const test = function({ a, b, c }) {
  return a + b + c;
};
module.exports = test;
```

The `valid-jsdoc` rule fails with the following message:

> Expected JSDoc for 'undefined' but found 'options'. (valid-jsdoc)

Changing the variable name in the `@param` tag to "undefined" doesn't work either.

ESLint 0.17.0

**Comment by nzakas** [comment]
Destructuring seems to confuse a lot of rules. Thanks for the report.

**Comment by lencioni** [comment]
Yeah, here I'm not even sure how one should properly document this in jsdoc.

**Comment by nzakas** [comment]
Working on this.

**Comment by robcolburn** [comment]
@nzakas,
The fix in a1da3b4 sort of dodges the issue a bit, and doesn't quite do what we'd hope it could…

It turns out JSDoc does support destructing via making up a placeholder name. It is lacking in official documentation.
https://github.com/jsdoc3/jsdoc/issues/987

```js
/**
 * @param {object} param
 * @param {number} param.a - First value
 * @param {object} param.b - Wrapper
 * @param {number} param.b.c - Second value
 * @return {number} sum a and b
 */
letters = ({a, b: {c}}) => a + c;
```

Perhaps, we can support this kind of grammar?

**Comment by mattyclarkson** [comment]
The following doesn't work either:

```
/**
 * @param {object} a - First value
 */
letters = ({a} = {}) => a;
```

**Comment by chrismcleod** [comment]
As pointed out by @mattyclarkson it seems eslint thinks a default value for a destructured param is a param itself and  gives the error "missing JSDoc for the parameter undefined"

![image](https://cloud.githubusercontent.com/assets/1134683/15130656/c6e396f2-160f-11e6-8157-391b8131e488.png)

**Comment by mysticatea** [comment]
I reported it recently: #6097

**Comment by aaronroberson** [comment]
It seems like this is a good behavior. Using destructuring in a method or function signature seems like a poor contract to make. 

I would suggest that it is a good practice to have all required items listed separately in the method signature and use a single parameter for a config object. 

Within the function body, it makes sense to use destructuring to pull out distinct properties from the provided config/options object. And it would make sense to use destructing when calling the method. But I don't think destructing belongs in the method signature itself.

**Comment by chrismcleod** [comment]
the reason i do it is to get something akin to ruby's named parameters.   Destructuring in the signature lets me refer to the parameters by name without having to reassign them or go through some other object.  If you use something like typescript, destructuring in the method signature allows you to type the properties of a **config** parameter without having to define a custom type.

**Comment by chrismcleod** [comment]
for example in typescript:
